### PR TITLE
.NET 8 OS compatibility test

### PR DESCRIPTION
**Pull request by ericsciple**
Based on feature flag state, performs a .NET 8 compatibility test during job-setup and optionally adds a warning if the test fails.

Also, reverted previous implementation:
- https://github.com/actions/runner/pull/3413

### CODE REVIEW

File: src/Runner.Worker/OSWarningChecker.cs [hunk 1]
```diff
@@ -80,8 +80,70 @@ public async Task CheckOSAsync(IExecutionContext context, IList<OSWarning> osWar
...
+                        arguments: string.Empty,
+                        environment: null,
+                        cancellationToken: context.CancellationToken);
+
+                    var outputStr = string.Join("\n", output).Trim();
+                    if (exitCode != 0 || !string.Equals(outputStr, "Hello from .NET 8!", StringComparison.Ordinal))
+                    {
+                        if (!omitAnnotation)
+                        {
+                            context.Warning("The runner is not compatible with .NET 8.");
```
Comment (ericsciple) [inline src/Runner.Worker/OSWarningChecker.cs:132]: Have the server send this message, which will solve the following problems:
- Act as a feature flag
- Can update from server to include a link to the changelog, etc.

File: src/Runner.Worker/OSWarningChecker.cs [hunk 2]
```diff
@@ -13,75 +13,77 @@ namespace GitHub.Runner.Worker
...
                             }
                         }
+                    };
+
+                    int exitCode = await process.ExecuteAsync(
+                        workingDirectory: HostContext.GetDirectory(WellKnownDirectory.Root),
+                        fileName: Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Bin), "testDotNet8Compatibility", $"TestDotNet8Compatibility{IOUtil.ExeExtension}"),
+                        arguments: string.Empty,
+                        environment: null,
+                        cancellationToken: context.CancellationToken);
```
Comment (TingluoHuang) [inline src/Runner.Worker/OSWarningChecker.cs:65]: i would consider set a static timeout about 10s, so the checking won't block the job forever when something went wrong...

File: src/Runner.Worker/OSWarningChecker.cs [hunk 3]
```diff
@@ -13,75 +13,77 @@ namespace GitHub.Runner.Worker
...
-                    Trace.Error("An error occurred while checking OS warnings for file '{0}' and regex '{1}'.", osWarning.FilePath, osWarning.RegularExpression);
-                    Trace.Error(ex);
-                    context.Global.JobTelemetry.Add(new JobTelemetry() { Type = JobTelemetryType.General, Message = $"An error occurred while checking OS warnings for file '{osWarning.FilePath}' and regex '{osWarning.RegularExpression}': {ex.Message}" });
-                }
+            }
+            catch (Exception ex)
+            {
+                Trace.Error("An error occurred while testing .NET 8 compatibility'");
+                Trace.Error(ex);
+                context.Global.JobTelemetry.Add(new JobTelemetry() { Type = JobTelemetryType.General, Message = $"An error occurred while testing .NET 8 compatibility; exception type '{ex.GetType().FullName}'; message: {ex.Message}" });
```
Comment (TingluoHuang) [inline src/Runner.Worker/OSWarningChecker.cs:85]: do we want to log the `output` into the telemetry, so we know what's failing?

Comment (TingluoHuang) [reply src/Runner.Worker/OSWarningChecker.cs:85]: maybe we got some output from the process, but the `process.ExecuteAsync` somehow throw later.

**Review by TingluoHuang — APPROVED** [review]

**Pull request merged**